### PR TITLE
Add Static Sha256 & Sha512 modules

### PR DESCRIPTION
**Pull request by tomijaga**
Hey @timohanke,
Thanks for the great work on this library. 

I had a use case where I needed the hasher to function properly and not loose its data after a canister upgrade.
Other people might have the same use-case as me, so i decided to update the code and make a PR.

Here are the changes I made:
- Created `StaticSha256` and `StaticSha512` types to hold the state required for calculating the hash
- Moved the s(0, 1, 2, ...) variables in a single array.
- Converted the methods for the Digest class to module functions in the Static modules

I updated the library in a way that is backward compatible with previous versions.
The import statements and methods were kept the same and a user can easily switch to the static version by adding **`/Static`** to the import statement like so:

```motoko
  import Sha256 "mo:sha2/Sha256/Static"
```

### CODE REVIEW

**Comment by timohanke** [comment]
Hi @tomijaga thanks for the PR. 

Would it have served your needs if the `Digest` class had a `share()` and `unshare()` function to get the internal state out and to reinstantiate it?

I would like to profile your PR to see if it has impact on instructions and heap allocations. But running `mops test` or `mops bench` fails for me with error
```
Canister <...> trapped explicitly: assertion failed at Static.mo:234.5-234.31, error code None
```
Does `mops test` work for you in your branch?

**Comment by tomijaga** [comment]
Hey @timohanke thanks for the response.

I couldn't find the bug causing the error in the `Sha512` module.
So, I commented it out and was able to run the benchmarks for `Sha256`.

| before | after |
| - | - | 
| image | image | 

The new code change increased the instructions by quite a bit.
This was a surprise to me, as I thought the static method performed better. Good to know.

Exporting the internal state via the `share()` and `unshare()` methods should be just fine for my use-case.
I will update the PR to reflect the changes.

**Comment by tomijaga** [comment]
I've updated the code and added tests for the new methods

**Comment by timohanke** [comment]
>I couldn't find the bug causing the error in the Sha512 module.

I fixed the bug here, just for completeness: https://github.com/research-ag/sha2/commit/b6e2dd798d9fb8e2775aae21c1d84cf7cfaca6bb

I could then run the full benchmark and had the same result: +60% in instructions for both sha256 and sha512.

I could also run my external profiling (not part of mops bench) to see the heap allocations. They are not affected by this PR and stay the same. 

So, yes, unless we find a way to improve instructions for the static version then probably share/unshare is the better approach. Does share/unshare work fine in your application?

**Comment by tomijaga** [comment]
Thanks for the fix, I see where i made the mistake now.


What I've done in the past to improve the static version is replace the record with a tuple where immutable fields are stored as entries and mutable fields are stored in a var array within the tuple. 
I updated the code here (https://github.com/NatLabs/sha2/commit/0efbcb915f186bd47c149bfb1b244b74230c639a) to compare the performance.

| record | tuple |
| - | - | 
| image | image |

The instructions for Sha256 went down from `+60%` to `46%` but Sha512 barely improved.

One drawback of the using the share/unshare is that you have to constantly export the state which creates some overhead.
The instructions for exporting the state will be equal to creating a new static record, seen in the first column with 0 blocks (`27_317`).
Adding this to the 1M blocks in the class version is still about `320M` instructions compared to the `460M` instructions in the static tuple version.
So even if I have to export the state after every 1M block, the class method performs better.

I am using the library to hash files, which will often be larger than 1MB, so share/unshare will work perfectly for my application.

**Comment by timohanke** [comment]
> One drawback of the using the share/unshare is that you have to constantly export the state which creates some overhead.

You don't do that once in preupgrade/postupgrade?

**Comment by timohanke** [comment]
Have you tried the class+ method? I mean where you have the `stable` declared global `StaticSha256` and pass that in through the constructor, and then class functions accesses it directly, not through a parameter. Don't know if that changes anything in terms of instructions, but was wondering if you tried. Then no share/unshare is needed and any intermediate state is automatically and instantly persisted without any additional share operation.

File: src/Sha256.mo [hunk 1]
```diff
@@ -211,6 +223,52 @@ module {
...
+        msg = Array.freeze(msg);
+        digest = Array.freeze(digest);
+
+        // state variables in Nat16 form
+        sh = [s0h, s1h, s2h, s3h, s4h, s5h, s6h, s7h];
+        sl = [s0l, s1l, s2l, s3l, s4l, s5l, s6l, s7l];
+      };
+
+      state;
+    };
```
Comment (AndyGura) [inline src/Sha256.mo:238]: Can be simplified like this:

```suggestion
    public func share() : StaticSha256  =  {
      i_msg; i_block; high; word;
      msg = Array.freeze(msg);
      digest = Array.freeze(digest);

      // state variables in Nat16 form
      sh = [s0h, s1h, s2h, s3h, s4h, s5h, s6h, s7h];
      sl = [s0l, s1l, s2l, s3l, s4l, s5l, s6l, s7l];
    };
```

File: src/Sha256.mo [hunk 2]
```diff
@@ -211,6 +223,52 @@ module {
...
+      i_msg := state.i_msg;
+      i_block := state.i_block;
+      high := state.high;
+      word := state.word;
+
+      var i = 0;
+      while (i < msg.size()){
+        msg[i] := state.msg[i];
+        i += 1;
+      };
```
Comment (AndyGura) [inline src/Sha256.mo:253]: Can be simplified:

```suggestion
      for (i in msg.keys()) {
        msg[i] := state.msg[i];
      };
```

File: src/Sha256.mo [hunk 3]
```diff
@@ -211,6 +223,52 @@ module {
...
+
+      i := 0;
+      while (i < digest.size()){
+        digest[i] := state.digest[i];
+        i += 1;
+      };
```
Comment (AndyGura) [inline src/Sha256.mo:259]: Same here
```suggestion
      for (i in digest.keys()) {
        digest[i] := state.digest[i];
      };
```

File: src/Sha512.mo [hunk 4]
```diff
@@ -222,6 +235,51 @@ module {
       };
     };
 
+    public func share() : StaticSha512 {
```
Comment (AndyGura) [inline src/Sha512.mo:238]: Same 3 code style suggestions as in Sha256.mo

File: test/sha2.test.mo [hunk 5]
```diff
@@ -281,3 +281,33 @@ do {
...
+
+do {
+  let data = Blob.fromArrayMut(Array.init<Nat8>(64, 0));
+  var digest256 = Sha256.Digest(#sha256);
+  var digest512 = Sha512.Digest(#sha512);
+
+  var digest256_state = digest256.share();
+  var digest512_state = digest512.share();
+
+  for (i in range(1,10000)) {
```
Comment (AndyGura) [inline test/sha2.test.mo:295]: Is there any particular reason to share/unshare data 10000 times in the test? Can we ensure that share/unshare work correctly by performing it once or twice?

**Review by AndyGura — CHANGES_REQUESTED** [review]
Hello @tomijaga ! Thank you for your contribution. I left few small comments in the PR, can you please check them?

I'm not sure why the test has 10k share/unshare calls, typically you would use `share` + `unshare` pair once per canister lifecycle between upgrades, e.g. `share`  to stable variable in the `preupgrade` hook and `unshare` from stable variable in the `postupgrade` hook. If it is for measuring instructions, we could add this to [sha2.bench.mo](https://github.com/research-ag/sha2/blob/main/bench/sha2.bench.mo), but I don't think it's significant

**Comment by timohanke** [comment]
We made a new PR https://github.com/research-ag/sha2/pull/7 based on this PR and merged it after some small edit. Thanks for providing it.

**Comment by timohanke** [comment]
The new version with share/unshare interface is now published as version 0.1.0 on mops https://mops.one/sha2

**Pull request closed without merging**